File: v3-sandbox/prototype/sdk/src/main/java/org/hiero/sdk/simple/network/keys/KeyPair.java
package org.hiero.sdk.simple.network.keys;

import java.util.Objects;
import org.jspecify.annotations.NonNull;

/**
 * Immutable tuple holding a {@link PrivateKey} and its corresponding {@link PublicKey}.
 *
 * @param privateKey the private key
 * @param publicKey  the corresponding public key
 */
public record KeyPair(@NonNull PrivateKey privateKey, @NonNull PublicKey publicKey) {

    /**
     * Validates that both record components are non-null.
     */
    public KeyPair {
        Objects.requireNonNull(privateKey, "Private key cannot be null");
        Objects.requireNonNull(publicKey, "Public key cannot be null");
    }

    /**
     * Creates a {@code KeyPair} from an existing {@link PrivateKey} by deriving the matching public key.
     *
     * @param privateKey the private key
     * @return a key pair consisting of {@code privateKey} and the derived public key
     */
    @NonNull
    public static KeyPair of(@NonNull final PrivateKey privateKey) {
        Objects.requireNonNull(privateKey, "Private key cannot be null");
        return new KeyPair(privateKey, privateKey.createPublicKey());
    }

    /**
     * Generates a fresh {@code KeyPair} for the given algorithm.
     *
     * @param algorithm the key algorithm
     * @return a new key pair
     */
    @NonNull
    public static KeyPair generate(@NonNull final KeyAlgorithm algorithm) {
        return of(PrivateKey.generate(algorithm));
    }
}

File: v3-sandbox/prototype-go/keys/key_pair.go
package keys

import "fmt"

// KeyPair holds a matching public and private key.
type KeyPair struct {
	publicKey  *PublicKey
	privateKey *PrivateKey
}

func (kp *KeyPair) PublicKey() *PublicKey   { return kp.publicKey }
func (kp *KeyPair) PrivateKey() *PrivateKey { return kp.privateKey }

// GenerateKeyPair creates a new random key pair for the given algorithm.
func GenerateKeyPair(algorithm KeyAlgorithm) (*KeyPair, error) {
	priv, err := GeneratePrivateKey(algorithm)
	if err != nil {
		return nil, fmt.Errorf("generate key pair: %w", err)
	}
	pub, err := priv.CreatePublicKey()
	if err != nil {
		return nil, fmt.Errorf("derive public key: %w", err)
	}
	return &KeyPair{publicKey: pub, privateKey: priv}, nil
}
